packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,1130 @@
+    if (index > 0) {
+      return <div
+        role='separator'
+        key={ key }
+        className={ css('ms-ComboBox-divider', styles.divider) } />;
+    }
+    return null;
+  }
+
+  private _renderHeader(item: IComboBoxOption): JSX.Element {
+    let { onRenderOption = this._onRenderOption } = this.props;
+    return (
+      <div key={ item.key } className={ css('ms-ComboBox-header', styles.header) } role='header'>
+        { onRenderOption(item, this._onRenderOption) }
+      </div>);
+  }
+
+  // Render menu item
+  @autobind
+  private _renderOption(item: IComboBoxOption): JSX.Element {
+    let { onRenderOption = this._onRenderOption } = this.props;
+    let id = this._id;
+    let isSelected: boolean = this._isOptionSelected(item.index);
+    return (
+      <CommandButton
+        id={ id + '-list' + item.index }
+        key={ item.key }
+        data-index={ item.index }
+        className={ css(
+          'ms-ComboBox-item', styles.item, {
+            ['is-selected ' + styles.itemIsSelected]:
+            isSelected,
+            ['is-disabled ' + styles.itemIsDisabled]: this.props.disabled === true
+          }
+        ) }
+        onClick={ () => this._onItemClick(item.index) }
+        role='option'
+        aria-selected={ isSelected ? 'true' : 'false' }
+        ariaLabel={ item.text }
+      > { <span ref={ this._resolveRef(isSelected ? '_selectedElement' : '') }>
+        { onRenderOption(item, this._onRenderOption) }
+      </span>
+        }
+      </CommandButton>
+    );
+  }
+
+  /**
+   * Use the current valid pending index if it exists OR
+   * we do not have a valid index and we currently have a pending input value,
+   * otherwise use the selected index
+   * */
+  @autobind
+  private _isOptionSelected(index: number): boolean {
+    let {
+      currentPendingValueValidIndex,
+      currentPendingValue,
+      selectedIndex
+    } = this.state;
+    return ((currentPendingValueValidIndex >= 0 || currentPendingValue !== '') ?
+      currentPendingValueValidIndex === index : selectedIndex === index);
+  }
+
+  /**
+   * Scroll the selected element into view
+   */
+  @autobind
+  private _scrollIntoView() {
+    if (this._selectedElement) {
+      let alignToTop = true;
+      if (this._comboBoxMenu.offsetParent) {
+        let scrollableParentRect = this._comboBoxMenu.offsetParent.getBoundingClientRect();
+        let selectedElementRect = this._selectedElement.offsetParent.getBoundingClientRect();
+
+        if (scrollableParentRect.top + scrollableParentRect.height <= selectedElementRect.top) {
+          alignToTop = false;
+        }
+      }
+
+      this._selectedElement.offsetParent.scrollIntoView(alignToTop);
+    }
+  }
+
+  // Render content of item
+  @autobind
+  private _onRenderOption(item: IComboBoxOption): JSX.Element {
+    return <span className={ css('ms-ComboBox-optionText', styles.optionText) } style={ { fontFamily: item.fontFamily } }>{ item.text }</span>;
```

Comment:
Wouldn't that mean that creators would have to pass in a component for the callback. What does having stateless react components get us?